### PR TITLE
added RepositoryDTO

### PR DESCRIPTION
**Pull request by plata**
new hierarchy level above CategoryDTO

required by #872

### CODE REVIEW

**Comment by qparis** [comment]
Please merge my pull request first so that you new DTOs are tested

File: phoenicis-repository/src/main/java/org/phoenicis/repository/dto/RepositoryDTO.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import java.util.*;
+
+/**
+ * Represents a repository
+ */
+@JsonDeserialize(builder = RepositoryDTO.Builder.class)
+public class RepositoryDTO {
+    private final String name;
+    private final List<CategoryDTO> categories;
+    private final Map<String, String> translations;
```
Comment (qparis) [inline phoenicis-repository/src/main/java/org/phoenicis/repository/dto/RepositoryDTO.java]: In general, we avoid using Map<String, String> in DTOs at it does not really say something about the data held.

Could we use a Set of TranslationDTO instead

File: phoenicis-repository/src/main/java/org/phoenicis/repository/dto/RepositoryDTO.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+import com.fasterxml.jackson.databind.annotation.JsonPOJOBuilder;
+import org.apache.commons.lang.builder.ToStringBuilder;
+
+import java.util.*;
+
+/**
+ * Represents a repository
+ */
+@JsonDeserialize(builder = RepositoryDTO.Builder.class)
+public class RepositoryDTO {
```
Comment (qparis) [inline phoenicis-repository/src/main/java/org/phoenicis/repository/dto/RepositoryDTO.java]: Please add equals and hashCode

File: phoenicis-repository/src/main/java/org/phoenicis/repository/repositoryTypes/ClasspathRepository.java [hunk 3]
```diff
@@ -63,10 +64,13 @@ public ClasspathRepository(String packagePath, ResourcePatternResolver resourceR
...
             Collections.sort(categoryDTOs, Comparator.comparing(CategoryDTO::getName));
-            return categoryDTOs;
+
+            final RepositoryDTO.Builder repositoryDTOBuilder = new RepositoryDTO.Builder()
+                    .withName("classpath repository").withCategories(categoryDTOs);
+            return repositoryDTOBuilder.build();
         } catch (IOException e) {
             LOGGER.warn("Error while reading resource directory", e);
-            return Collections.emptyList();
+            return null;
```
Comment (qparis) [inline phoenicis-repository/src/main/java/org/phoenicis/repository/repositoryTypes/ClasspathRepository.java]: You should return an empty RepositoryDTO here

File: phoenicis-repository/src/main/java/org/phoenicis/repository/repositoryTypes/GitRepository.java [hunk 4]
```diff
@@ -71,11 +67,11 @@ public GitRepository(URI gitRepositoryUri, String cacheDirectoryPath,
                 LOGGER.error(String.format("Couldn't create folder for git repository '%s' at '%s'",
                         this.gitRepositoryUri, gitRepositoryLocation.getAbsolutePath()));
 
-                return Collections.emptyList();
+                return null;
```
Comment (qparis) [inline phoenicis-repository/src/main/java/org/phoenicis/repository/repositoryTypes/GitRepository.java]: Please return an empty RepositoryDTO here

File: phoenicis-repository/src/main/java/org/phoenicis/repository/repositoryTypes/LocalRepository.java [hunk 5]
```diff
@@ -60,15 +57,34 @@ private LocalRepository(File repositoryDirectory, ObjectMapper objectMapper) {
...
     @Override
-    public List<CategoryDTO> fetchInstallableApplications() {
+    public RepositoryDTO fetchInstallableApplications() {
+
         final File[] categoryDirectories = repositoryDirectory.listFiles();
 
         if (categoryDirectories == null) {
-            return Collections.emptyList();
+            return null;
         }
```
Comment (qparis) [inline phoenicis-repository/src/main/java/org/phoenicis/repository/repositoryTypes/LocalRepository.java]: Please return an empty RepositoryDTO here

File: phoenicis-repository/src/main/java/org/phoenicis/repository/repositoryTypes/NullRepository.java [hunk 6]
```diff
@@ -19,15 +19,12 @@
...
-import java.util.Collections;
-import java.util.List;
+import org.phoenicis.repository.dto.RepositoryDTO;
 
 public class NullRepository implements Repository {
     @Override
-    public List<CategoryDTO> fetchInstallableApplications() {
-        return Collections.emptyList();
+    public RepositoryDTO fetchInstallableApplications() {
+        return null;
```
Comment (qparis) [inline phoenicis-repository/src/main/java/org/phoenicis/repository/repositoryTypes/NullRepository.java]: Same here

File: phoenicis-repository/src/test/java/org/phoenicis/repository/repositoryTypes/NullRepositoryTest.java [hunk 7]
```diff
@@ -19,13 +19,12 @@
...
 import org.junit.Test;
-import org.phoenicis.repository.repositoryTypes.NullRepository;
 
 import static org.junit.Assert.assertEquals;
 
 public class NullRepositoryTest {
     @Test
     public void testNullRepositoryTest() {
-        assertEquals(0, new NullRepository().fetchInstallableApplications().size());
+        assertEquals(null, new NullRepository().fetchInstallableApplications());
```
Comment (qparis) [inline phoenicis-repository/src/test/java/org/phoenicis/repository/repositoryTypes/NullRepositoryTest.java]: You may want to use assertNull

File: phoenicis-tests/src/main/java/org/phoenicis/tests/MockedRepository.java [hunk 8]
```diff
@@ -38,17 +38,21 @@
...
-                                                            OperatingSystem.MACOSX))
-                                                    .withScript(wineScript()).build()))
-                                            .build()))
-                                    .withName("Functions").build());
+        public RepositoryDTO fetchInstallableApplications() {
+            return new RepositoryDTO.Builder()
+                    .withCategories(
+                            Collections.singletonList(
+                                    new CategoryDTO.Builder()
+                                            .withApplications(Arrays.asList(new ApplicationDTO.Builder().withName(
```
Comment (qparis) [inline phoenicis-tests/src/main/java/org/phoenicis/tests/MockedRepository.java]: The formating here is strange

**Comment by qparis** [comment]
Please merge the master to this branch to run DTO tests

File: phoenicis-tests/src/main/java/org/phoenicis/tests/MockedRepository.java (hunk 8 above)
Comment (plata) [inline phoenicis-tests/src/main/java/org/phoenicis/tests/MockedRepository.java]: I cannot change it. The formatter does it this way.

Comment (qparis) [inline phoenicis-tests/src/main/java/org/phoenicis/tests/MockedRepository.java]: Then the formatter needs to be fixed or we need to remove it. 
The point of having code conventions is to make it as more readable as possible

**Comment by qparis** [comment]
With your solution, what if we have two scripts that have the same name in two different repositories?

File: phoenicis-tests/src/main/java/org/phoenicis/tests/MockedRepository.java (hunk 8 above)
Comment (plata) [inline phoenicis-tests/src/main/java/org/phoenicis/tests/MockedRepository.java]: I guess in this particular case, we should build the objects step by step and not in one big call. Anyways this is completely unrelated to the PR.

**Comment by plata** [comment]
The merge process should not have changed, so I would say: same behavior as before.

File: phoenicis-tests/src/main/java/org/phoenicis/tests/MockedRepository.java (hunk 8 above)
Comment (qparis) [inline phoenicis-tests/src/main/java/org/phoenicis/tests/MockedRepository.java]: One advantage of the builder pattern is to avoid building the objects step by steps. 
Code formatting should help us. If it is not the case, we should get rid of it. 
This is not the only place where it causes problems (see Spring configuration)
Agree that it is not related to this PR, but we should track this issue though
